state/address.go

```diff
@@ -88,57 +89,161 @@ func (st *State) Addresses() ([]string, error) {
+		if err != nil {
+			return nil, errors.Trace(err)
 		}
-		hostPorts := networkHostsPorts(existingDoc.APIHostPorts)
-		if !hostsPortsEqual(netHostsPorts, hostPorts) {
-			op.Update = bson.D{{
-				"$set", bson.D{{"apihostports", doc.APIHostPorts}},
-			}}
-		} else {
+		ops = append(ops, agentAddrOps...)
+
+		if ops == nil || len(ops) == 0 {
 			return nil, statetxn.ErrNoOperations
 		}
-		return []txn.Op{op}, nil
+		return ops, nil
 	}
+
 	if err := st.db().Run(buildTxn); err != nil {
 		return errors.Annotate(err, "cannot set API addresses")
 	}
-	logger.Debugf("setting API hostPorts: %v", netHostsPorts)
 	return nil
 }
 
-// APIHostPorts returns the API addresses as set by SetAPIHostPorts.
+// getOpsForHostPortsChange returns a slice of operations used to update an
+// API host/collection in the DB.
+// If the current document indicates the same host/port collection as the
+// input, no operations are returned.
+func (st *State) getOpsForHostPortsChange(
+	mc mongo.Collection,
+	key string,
+	newHostPorts [][]network.HostPort,
+) ([]txn.Op, error) {
+	var ops []txn.Op
+
+	// Retrieve the current document. Return an insert operation if not found.
+	var extantHostPortDoc apiHostPortsDoc
+	err := mc.Find(bson.D{{"_id", key}}).One(&extantHostPortDoc)
+	if err != nil {
+		if err == mgo.ErrNotFound {
+			return []txn.Op{{
+				C:      controllersC,
+				Id:     key,
+				Insert: bson.D{{"apihostports", fromNetworkHostsPorts(newHostPorts)}},
+			}}, nil
+		}
+		return ops, err
+	}
+
+	// Queue an update operation if the host/port collections differ.
+	extantHostPorts := networkHostsPorts(extantHostPortDoc.APIHostPorts)
+	if !hostsPortsEqual(newHostPorts, extantHostPorts) {
+		ops = []txn.Op{{
+			C:  controllersC,
+			Id: key,
+			Assert: bson.D{{
+				"txn-revno", extantHostPortDoc.TxnRevno,
+			}},
+			Update: bson.D{{
+				"$set", bson.D{{"apihostports", fromNetworkHostsPorts(newHostPorts)}},
+			}},
+		}}
+		logger.Debugf("setting %s: %v", key, newHostPorts)
+	}
+	return ops, nil
+}
+
+// filterHostPortsForManagementSpace filters the collection of API addresses
+// based on the configured management space for the controller.
+// If there is no space configured, or if the one of the slices is filtered down
+// to zero elements, just use the unfiltered slice for safety - we do not
+// want to cut off communication to the controller based on erroneous config.
+func (st *State) filterHostPortsForManagementSpace(apiHostPorts [][]network.HostPort) ([][]network.HostPort, error) {
+	config, err := st.ControllerConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	var hostPortsForAgents [][]network.HostPort
+	if mgmtSpace := config.JujuManagementSpace(); mgmtSpace != "" {
+		hostPortsForAgents = make([][]network.HostPort, len(apiHostPorts))
+		sp := network.SpaceName(mgmtSpace)
+		for i := range apiHostPorts {
+			if filtered, ok := network.SelectHostsPortBySpaces(apiHostPorts[i], sp); ok {
+				hostPortsForAgents[i] = filtered
```

Comment:
what happens if filtered is empty?

Reply:
The guard for this is in SelectHostsPortBySpaces. It will return false if the list is empty.